core/src/main/scala/scalaz/FreeT.scala

```diff
@@ -39,14 +50,50 @@ sealed abstract class FreeT[S[_], M[_], A] {
   final def map[B](f: A => B)(implicit S: Functor[S], M: Functor[M]): FreeT[S, M, B] =
     this match {
       case Suspend(m) => Suspend(M.map(m)(_.bimap(f, S.lift(_.map(f)))))
-      case Gosub(a, k) => Gosub(a, (x: Any) => k(x).map(f))
+      case g @ Gosub() => gosub(g.a)(g.f.andThen(_.map(f)))
     }
 
   /** Binds the given continuation to the result of this computation. */
   final def flatMap[B](f: A => FreeT[S, M, B]): FreeT[S, M, B] =
     this match {
-      case Gosub(a, k) => Gosub(a, (x: Any) => Gosub(k(x), f))
-      case a => Gosub(a, f)
+      case g @ Gosub() => gosub(g.a)(x => gosub(g.f(x))(f))
+      case a => gosub(a)(f)
+    }
+
+  /** Same as `hoistM` but different constraints */
+  def hoistN[N[_]](mn: M ~> N)(implicit S: Functor[S], N: Functor[N]): FreeT[S, N, A] =
```

Comment:
Can we improve this doc? The purescript doc says `Change the underlying `Monad` for a `FreeT` action.` and the signature here is not as obvious as it is an instance method. Perhaps something like:

    /**
     * Changes the `Monad` for this `FreeT`, ie. turning
     * this `FreeT[S, M, A]` into a `FreeT[S, N, A]`
     * given Functors for `S` and `N`
     */